### ISSUE TITLE
[BUG] `useViewportScroll` is not updating.

### ISSUE THREAD

**Issue by saisandeepvaddi** [framer/motion#835]
**1. Read the [FAQs](#faqs) 👇**

**2. Describe the bug**

Give a clear and concise description of what the bug is.
`scrollYProgress` from `useViewportScroll` hook does not update.

**3. IMPORTANT: Provide a CodeSandbox reproduction of the bug**

A CodeSandbox minimal reproduction will allow us to quickly follow the reproduction steps. **Without one, this bug report won't be accepted.**

**4. Steps to reproduce**

1. Go to https://codesandbox.io/s/trusting-jepsen-mikfs?file=/src/App.js
2. Open console
3. Scroll down
4. No updated `scrollYProgress` value

**5. Expected behavior**

`scrollYProgress` updates when scroll happens.

**Comment by mattgperry** [comment]
Do you have body/html set to height: 100%?

**Comment by saisandeepvaddi** [comment]
Nope.

This is the whole css

```css
body {
  padding: 0;
  margin: 0;
  display: flex;
  justify-content: center;
  align-items: center;
}

.App {
  font-family: sans-serif;
  text-align: center;
}

.grid {
  display: flex;
  flex-wrap: wrap;
  height: 800px;
}

.box {
  margin: 10;
}


```

**Comment by camsloanftc** [comment]
The scroll progress happens outside of the react render loop (I believe it is for performance reasons), so you actually need something like this:

```js
  useEffect(() => {
    return scrollYProgress.onChange((v) => console.log(v));
  }, [scrollYProgress]);

```

I tweaked your demo [code example](https://codesandbox.io/s/infallible-andras-t33nx).

**Comment by paales** [comment]
useViewportScroll is a `MotionValue<number>` To fix your example, you should use `useMotionTemplate`

Example:
https://codesandbox.io/s/musing-carson-vyt4i?file=/src/App.js:874-882

**Comment by saisandeepvaddi** [comment]
@camsloanftc In your example, in the console.log, progress doesn't seem to be going to 100% even when scrolled to the bottom right?

**Comment by saisandeepvaddi** [comment]
@paales Your example seems to be updating the values, but, the code in my initial example is just the same as the one in the docs right?

Docs say

```
export const MyComponent = () => {
  const { scrollYProgress } = useViewportScroll()
  return <motion.div style={{ scaleX: scrollYProgress }} />
}

```

Regardless of the usage of `useEffect` in my example, using `scaleX` with `scrollYProgress` must be doing something based on the docs. I may be missing something.

**Comment by FernandoJMedina** [comment]
Same happens to me.

**Comment by FernandoJMedina** [comment]
```
- Delete the 800px height of the grid class.
- Add right: 0 to your div bar.
 const scaleX = useTransform(scrollYProgress , [0,1], [0,1])
- Add scaleX variable to the css property at your div.motion element
```

**Comment by FernandoJMedina** [comment]
Delete 800px height on css grid class.
You must give your motion.div a width of '100%' or apply right: 0.
Then you have to use useTransform hook to interpolate the scrollYprogress motionvalue.

`const scaleX = useTransform(scrollYProgress, [0, 1], [1, 0]);`

Then apply the variable to de scaleX css property at your motion.div

**Comment by ikupenov** [comment]
I ran into this couple of days ago and the issue stems from the fact that I had set the height to 100vh/100%. Try running the following snippet in your browser: `document.body.clientHeight - window.innerHeight <= 0`. If it returns true you're probably running into the same issue and will need to use min-height rather than a fixed height for the body element.

**Comment by Eduardo-Balistieri** [comment]
You must specify a width to apply the scale.
And regarding the scale, I don’t understand the reason for the negative value.

```javascript
<motion.div
  style={{
    position: "fixed",
    top: 0,
    left: 0,
    width: '100%',
    height: 10,
    scaleX: scrollYProgress,
    background: "red",
  }}
></motion.div>
```

**Comment by brEaTh-263** [comment]
> I ran into this couple of days ago and the issue stems from the fact that I had set the height to 100vh/100%. Try running the following snippet in your browser: `document.body.clientHeight - window.innerHeight <= 0`. If it returns true you're probably running into the same issue and will need to use min-height rather than a fixed height for the body element.

Any workaround @ikupenov if one is using scroll snap as well? Because without 100vh height scroll snap breaks..

**Comment by RohaanAnsari** [comment]
> I ran into this couple of days ago and the issue stems from the fact that I had set the height to 100vh/100%. Try running the following snippet in your browser: `document.body.clientHeight - window.innerHeight <= 0`. If it returns true you're probably running into the same issue and will need to use min-height rather than a fixed height for the body element

**Comment by mattgperry** [comment]
Yeah this is usually the "height" of the whole page has been set to something equal to or smaller than the document, even when the page still scrolls

**Comment by johanjmoncada** [comment]
I had to do an infinite scroll and request content when the current scroll position was at the middle of the track. 

Since I had to update the state each time I got a new item from API, I had to `setItems` and what it happened, the `scrollYProgress.onChange` was re-instance  per each `setItems`. It was leading me to performance issues and weird behaviour.

How I resolved it was doing:

```
const [items, setItems] = useState<ObjectItem[]>([InitialItemsProps])

const scrollHandler = async (scrollPositionPercent: number) => {
  console.log(scrollPositionPercent)
  if (scrollPositionPercent > 0.5) {
      // Request API to get more results
      const requestedItem = { ... }
      setItems((currentItems) => [
        ...currentItems,
        requestedItem as ObjectItem,
      ])
    }
  }
}

 useEffect(() => {
    scrollYProgress.onChange(scrollHandler)
  }, [])
```

The key part of that was:
- Adding the part `scrollYProgress.onChange(scrollHandler)` inside a useEffect that runs only when the component is mount
- `setItems` is receiving a callback, where its props is the current state.

I dunno if this is the best solution, but after spending some hours, it's working and I'm happy. Hope that this small comment helps someone else out there.
